cohorts/2023/02-experiment-tracking/homework/train_pipeline.py

```diff
@@ -0,0 +1,241 @@
+            )
+
+            mlflow.log_metric("RMSE_Val", rmse_val)
+            logger.info("Trial %s: RMSE_Val = %s", trial.number, rmse_val)
+
+            mlflow.log_metric("RMSE_Test", rmse_test)
+            logger.info("Trial %s: RMSE_Test = %s", trial.number, rmse_test)
+
+            model_id = mlflow_exp_obj.get_max_model_id(hpo_run_id)
+
+            mlflow.set_tag("model_id", str(model_id))
+
+            mlflow.set_tag(
+                "data_info", str({k: v["month"] for k, v in file_names.items()})
+            )
+
+            # log pipeline as an mlflow artifact
+            mlflow.sklearn.log_model(sk_model=pipeline_opt, artifact_path="model")
+
+        return rmse_test
+
+    study = optuna.create_study(direction="minimize", sampler=TPESampler(seed=42))
+    study.optimize(objective, n_trials=num_trials)
+
+
+if __name__ == "__main__":
+    # Get absolute path of this module and its directory path
+    FILE_PATH = Path(__file__).resolve()
+    BASE_PATH = FILE_PATH.parent
+
+    config_path = os.path.join(BASE_PATH, "config.yaml")
+
+    with open(config_path, "r", encoding="utf-8") as filename:
+        config = yaml.safe_load(filename)
+
+    # attributing required constants
+    DATA_DIR = BASE_PATH / "data"
+    S3_URL = config["constants"]["S3_URL"]
+    PREFIX = config["constants"]["PREFIX"]
+    HPO_EXPERIMENT_NAME = config["constants"]["HPO_EXPERIMENT_NAME"]
+    HPO_BEST_MODEL = config["constants"]["HPO_BEST_MODEL"]
+    HPO_CHAMPION_MODEL = config["constants"]["HPO_CHAMPION_MODEL"]
+
+    parser = argparse.ArgumentParser(description="MLOPS Zoomcamp Homework 2")
+    parser.add_argument(
+        "--train-month",
+        type=str,
+        required=True,
+        help="Train month in format YYYY-MM",
+    )
+    parser.add_argument(
+        "--val-month",
+        type=str,
+        required=True,
+        help="Validation month in format YYYY-MM",
+    )
+    parser.add_argument(
+        "--test-month",
+        type=str,
+        required=True,
+        help="Test month in format YYYY-MM",
+    )
+    parser.add_argument(
+        "--num-trials",
+        type=int,
+        default=2,
+        help="Number of trials for each Run at mlflow",
+    )
+    parser.add_argument(
+        "--flag-reset-mlflow",
+        type=str,
+        default="N",
+        help="Flag to ask mlflow reset - deleting DB and mlruns artifacts folders. "
+        "Y=reset, N=no reset",
+    )
+    args = parser.parse_args()
+
+    FILE_NAMES = {
+        "train": {"month": args.train_month},
+        "val": {"month": args.val_month},
+        "test": {"month": args.test_month},
+    }
+
+    # Instantiate the DataHandler object
+    data_handler = DataHandler(DATA_DIR, S3_URL)
+
+    # Download the train, validation, and test data and update the dictionary
+    # with corresponding file_paths required for splitting datasets function
+    for file_type, month_info in FILE_NAMES.items():
```

Comment:
I would personally change most of the functionalities of this "if __name__=='__main__': to different individual functions and then call them in this block, it might be hard to understand the code when all in the same location. Ex: (def create_parser: ... | def download_data: ... ) etc